src/machine.rs

```diff
@@ -205,6 +205,23 @@ impl interpret::Provenance for Provenance {
             Provenance::Wildcard => None,
         }
     }
+
+    fn join(left: Option<Self>, right: Option<Self>) -> Option<Self> {
+        match (left, right) {
+            // If one side is `None`, the result is `None`.
+            (None, _) | (_, None) => None,
```

Comment:
At that point maybe just use

```suggestion
        // If one side is `None`, the result is `None`.
        match (left?, right?) {
```

Reply:
However, I just realized I can remove this first match arm entirely, and behavior does not even change.